File: packages/runar-py/runar/sdk/anf_interpreter.py
"""Lightweight ANF interpreter for auto-computing state transitions.

Given a compiled artifact's ANF IR, the current contract state, and
method arguments, this interpreter walks the ANF bindings and computes
the new state. It handles ``update_prop`` nodes to track state mutations,
while skipping on-chain-only operations like ``check_preimage``,
``deserialize_state``, ``get_state_script``, ``add_output``, and ``add_raw_output``.

This enables the SDK to auto-compute ``newState`` for stateful contract
calls, so callers don't need to duplicate contract logic.
"""

from __future__ import annotations

import hashlib
import re
from typing import Any, Dict, List, Optional


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def compute_new_state(
    anf: dict,
    method_name: str,
    current_state: dict,
    args: dict,
) -> dict:
    """Compute the new state after executing a contract method.

    Args:
        anf: The ANF IR from the compiled artifact (plain dict from JSON).
        method_name: The method to execute (must be a public method).
        current_state: Current contract state (property name -> value).
        args: Method arguments (param name -> value).

    Returns:
        The updated state (merged with current_state).
    """
    method = None
    for m in anf.get('methods', []):
        if m['name'] == method_name and m.get('isPublic', False):
            method = m
            break

    if method is None:
        raise ValueError(
            f"computeNewState: method '{method_name}' not found in ANF IR"
        )

    # Initialize the environment with property values and method params
    env: Dict[str, Any] = {}

    # Load properties
    for prop in anf.get('properties', []):
        name = prop['name']
        env[name] = current_state.get(name, prop.get('initialValue'))

    # Load method params (skip implicit ones injected by the compiler)
    implicit_params = {'_changePKH', '_changeAmount', '_newAmount', 'txPreimage'}
    for param in method.get('params', []):
        pname = param['name']
        if pname in implicit_params:
            continue
        if pname in args:
            env[pname] = args[pname]

    # Track state mutations
    state_delta: Dict[str, Any] = {}

    # Walk bindings
    _eval_bindings(method.get('body', []), env, state_delta, anf)

    return {**current_state, **state_delta}


# ---------------------------------------------------------------------------
# Binding evaluation
# ---------------------------------------------------------------------------

def _eval_bindings(
    bindings: List[dict],
    env: Dict[str, Any],
    state_delta: Dict[str, Any],
    anf: Optional[dict] = None,
) -> None:
    for binding in bindings:
        val = _eval_value(binding['value'], env, state_delta, anf)
        env[binding['name']] = val


def _eval_value(
    value: dict,
    env: Dict[str, Any],
    state_delta: Dict[str, Any],
    anf: Optional[dict] = None,
) -> Any:
    kind = value.get('kind', '')

    if kind == 'load_param':
        return env.get(value['name'])

    if kind == 'load_prop':
        return env.get(value['name'])

    if kind == 'load_const':
        v = value.get('value')
        # Handle @ref: aliases
        if isinstance(v, str) and v.startswith('@ref:'):
            return env.get(v[5:])
        return v

    if kind == 'bin_op':
        return _eval_bin_op(
            value['op'],
            env.get(value['left']),
            env.get(value['right']),
            value.get('result_type'),
        )

    if kind == 'unary_op':
        return _eval_unary_op(
            value['op'],
            env.get(value['operand']),
            value.get('result_type'),
        )

    if kind == 'call':
        call_args = [env.get(a) for a in value.get('args', [])]
        return _eval_call(value['func'], call_args)

    if kind == 'method_call':
        call_args = [env.get(a) for a in value.get('args', [])]
        return _eval_method_call(env.get(value.get('object')), value.get('method'), call_args, env, state_delta, anf)

    if kind == 'if':
        cond = env.get(value['cond'])
        branch = value['then'] if _is_truthy(cond) else value['else']
        child_env = dict(env)
        _eval_bindings(branch, child_env, state_delta, anf)
        env.update(child_env)
        if branch:
            return child_env.get(branch[-1]['name'])
        return None

    if kind == 'loop':
        count = value.get('count', 0)
        body = value.get('body', [])
        iter_var = value.get('iterVar', '')
        last_val = None
        for i in range(count):
            env[iter_var] = i
            loop_env = dict(env)
            _eval_bindings(body, loop_env, state_delta, anf)
            env.update(loop_env)
            if body:
                last_val = loop_env.get(body[-1]['name'])
        return last_val

    if kind == 'assert':
        return None

    if kind == 'update_prop':
        new_val = env.get(value['value'])
        env[value['name']] = new_val
        state_delta[value['name']] = new_val
        return None

    # add_output -- process stateValues to update mutable properties
    if kind == 'add_output':
        state_values = value.get('stateValues', [])
        if state_values and anf:
            mutable_props = [
                p['name'] for p in anf.get('properties', [])
                if not p.get('readonly', False)
            ]
            for i, sv in enumerate(state_values):
                if i < len(mutable_props):
                    resolved = env.get(sv)
                    prop_name = mutable_props[i]
                    env[prop_name] = resolved
                    state_delta[prop_name] = resolved
        return None

    # On-chain-only operations -- skip in simulation
    if kind in ('check_preimage', 'deserialize_state', 'get_state_script',
                'add_raw_output'):
        return None

    return None


# ---------------------------------------------------------------------------
# Binary operations
# ---------------------------------------------------------------------------

def _eval_bin_op(op: str, left: Any, right: Any, result_type: Optional[str] = None) -> Any:
    if result_type == 'bytes' or (isinstance(left, str) and isinstance(right, str)):
        return _eval_bytes_bin_op(op, str(left or ''), str(right or ''))

    l = _to_int(left)
    r = _to_int(right)

    if op == '+':
        return l + r
    if op == '-':
        return l - r
    if op == '*':
        return l * r
    if op == '/':
        return 0 if r == 0 else _truncate_div(l, r)
    if op == '%':
        return 0 if r == 0 else _truncate_mod(l, r)
    if op in ('==', '==='):
        return l == r
    if op in ('!=', '!=='):
        return l != r
    if op == '<':
        return l < r
    if op == '<=':
        return l <= r
    if op == '>':
        return l > r
    if op == '>=':
        return l >= r
    if op in ('&&', 'and'):
        return _is_truthy(left) and _is_truthy(right)
    if op in ('||', 'or'):
        return _is_truthy(left) or _is_truthy(right)
    if op == '&':
        return l & r
    if op == '|':
        return l | r
    if op == '^':
        return l ^ r
    if op == '<<':
        return l << r
    if op == '>>':
        return l >> r
    return 0


def _truncate_div(a: int, b: int) -> int:
    """Integer division truncating toward zero (matching JS/Bitcoin semantics)."""
    return int(a / b) if (a < 0) != (b < 0) and a % b != 0 else a // b


def _truncate_mod(a: int, b: int) -> int:
    """Modulo matching truncation toward zero."""
    return a - _truncate_div(a, b) * b


def _eval_bytes_bin_op(op: str, left: str, right: str) -> Any:
    if op == '+':  # cat
        return left + right
    if op in ('==', '==='):
        return left == right
    if op in ('!=', '!=='):
        return left != right
    return ''


# ---------------------------------------------------------------------------
# Unary operations
# ---------------------------------------------------------------------------

def _eval_unary_op(op: str, operand: Any, result_type: Optional[str] = None) -> Any:
    if result_type == 'bytes':
        if op == '~':
            hex_str = str(operand or '')
            b = bytes.fromhex(hex_str)
            return bytes(~x & 0xff for x in b).hex()
        return operand

    val = _to_int(operand)
    if op == '-':
        return -val
    if op in ('!', 'not'):
        return not _is_truthy(operand)
    if op == '~':
        return ~val
    return val


# ---------------------------------------------------------------------------
# Built-in function calls
# ---------------------------------------------------------------------------

def _eval_call(func: str, args: List[Any]) -> Any:
    # Crypto -- mock
    if func in ('checkSig', 'checkMultiSig', 'checkPreimage'):
        return True

    # Crypto -- real hashes
    if func == 'sha256':
        return _hash_fn('sha256', args[0])
    if func == 'hash256':
        return _hash_fn('hash256', args[0])
    if func == 'hash160':
        return _hash_fn('hash160', args[0])
    if func == 'ripemd160':
        return _hash_fn('ripemd160', args[0])

    # Assert -- skip
    if func == 'assert':
        return None

    # Byte operations
    if func == 'num2bin':
        n = _to_int(args[0])
        length = int(_to_int(args[1]))
        return _num2bin_hex(n, length)
    if func == 'bin2num':
        return _bin2num_int(str(args[0] or ''))
    if func == 'cat':
        return str(args[0] or '') + str(args[1] or '')
    if func == 'substr':
        hex_str = str(args[0] or '')
        start = int(_to_int(args[1]))
        length = int(_to_int(args[2]))
        return hex_str[start * 2:(start + length) * 2]
    if func == 'reverseBytes':
        hex_str = str(args[0] or '')
        pairs = [hex_str[i:i + 2] for i in range(0, len(hex_str), 2)]
        return ''.join(reversed(pairs))
    if func == 'len':
        hex_str = str(args[0] or '')
        return len(hex_str) // 2

    # Math builtins
    if func == 'abs':
        return abs(_to_int(args[0]))
    if func == 'min':
        return min(_to_int(args[0]), _to_int(args[1]))
    if func == 'max':
        return max(_to_int(args[0]), _to_int(args[1]))
    if func == 'within':
        x = _to_int(args[0])
        return x >= _to_int(args[1]) and x < _to_int(args[2])
    if func == 'safediv':
        d = _to_int(args[1])
        return 0 if d == 0 else _truncate_div(_to_int(args[0]), d)
    if func == 'safemod':
        d = _to_int(args[1])
        return 0 if d == 0 else _truncate_mod(_to_int(args[0]), d)
    if func == 'clamp':
        v, lo, hi = _to_int(args[0]), _to_int(args[1]), _to_int(args[2])
        return lo if v < lo else hi if v > hi else v
    if func == 'sign':
        v = _to_int(args[0])
        return 1 if v > 0 else -1 if v < 0 else 0
    if func == 'pow':
        base = _to_int(args[0])
        exp = _to_int(args[1])
        if exp < 0:
            return 0
        return base ** exp
    if func == 'sqrt':
        v = _to_int(args[0])
        if v <= 0:
            return 0
        x = v
        y = (x + 1) // 2
        while y < x:
            x = y
            y = (x + v // x) // 2
        return x
    if func == 'gcd':
        a, b = abs(_to_int(args[0])), abs(_to_int(args[1]))
        while b:
            a, b = b, a % b
        return a
    if func == 'divmod':
        a = _to_int(args[0])
        b = _to_int(args[1])
        if b == 0:
            return 0
        return _truncate_div(a, b)
    if func == 'log2':
        v = _to_int(args[0])
        if v <= 0:
            return 0
        bits = 0
        x = v
        while x > 1:
            x >>= 1
            bits += 1
        return bits
    if func == 'bool':
        return 1 if _is_truthy(args[0]) else 0
    if func == 'mulDiv':
        return _truncate_div(_to_int(args[0]) * _to_int(args[1]), _to_int(args[2]))
    if func == 'percentOf':
        return _truncate_div(_to_int(args[0]) * _to_int(args[1]), 10000)

    # Preimage intrinsics -- return dummy values in simulation
    if func in ('extractOutputHash', 'extractAmount'):
        return '00' * 32

    return None


def _eval_method_call(
    obj: Any,
    method: Optional[str],
    args: List[Any],
    caller_env: Optional[Dict[str, Any]] = None,
    state_delta: Optional[Dict[str, Any]] = None,
    anf: Optional[dict] = None,
) -> Any:
    # Look up private method in ANF IR
    if anf and method:
        for m in anf.get('methods', []):
            if m['name'] == method and not m.get('isPublic', False):
                # Create new env with property values from caller
                new_env: Dict[str, Any] = {}
                if caller_env:
                    for prop in anf.get('properties', []):
                        name = prop['name']
                        if name in caller_env:
                            new_env[name] = caller_env[name]
                # Map method params to passed args
                params = m.get('params', [])
                for i, param in enumerate(params):
                    if i < len(args):
                        new_env[param['name']] = args[i]
                # Evaluate method body
                body = m.get('body', [])
                child_delta: Dict[str, Any] = {}
                _eval_bindings(body, new_env, child_delta, anf)
                # Propagate state delta back
                if state_delta is not None:
                    state_delta.update(child_delta)
                if caller_env is not None:
                    for k, v in child_delta.items():
                        caller_env[k] = v
                # Return last binding's value
                if body:
                    return new_env.get(body[-1]['name'])
                return None
    return None


# ---------------------------------------------------------------------------
# Hash helpers
# ---------------------------------------------------------------------------

def _hash_fn(name: str, input_val: Any) -> str:
    hex_str = str(input_val or '')
    data = bytes.fromhex(hex_str)

    if name == 'sha256':
        return hashlib.sha256(data).hexdigest()
    if name == 'hash256':
        return hashlib.sha256(hashlib.sha256(data).digest()).hexdigest()
    if name == 'ripemd160':
        return hashlib.new('ripemd160', data).hexdigest()
    if name == 'hash160':
        return hashlib.new('ripemd160', hashlib.sha256(data).digest()).hexdigest()
    return ''


# ---------------------------------------------------------------------------
# Numeric helpers
# ---------------------------------------------------------------------------

_BIGINT_RE = re.compile(r'^-?\d+n$')
_INT_RE = re.compile(r'^-?\d+$')


def _to_int(v: Any) -> int:
    if isinstance(v, int) and not isinstance(v, bool):
        return v
    if isinstance(v, bool):
        return 1 if v else 0
    if isinstance(v, float):
        return int(v)
    if isinstance(v, str):
        # Handle "42n" format from JSON
        if _BIGINT_RE.match(v):
            return int(v[:-1])
        if _INT_RE.match(v):
            return int(v)
        return 0
    return 0


def _is_truthy(v: Any) -> bool:
    if isinstance(v, bool):
        return v
    if isinstance(v, int):
        return v != 0
    if isinstance(v, float):
        return v != 0.0
    if isinstance(v, str):
        return v != '' and v != '0' and v != 'false'
    return False


# ---------------------------------------------------------------------------
# Byte encoding helpers
# ---------------------------------------------------------------------------

def _num2bin_hex(n: int, byte_len: int) -> str:
    if n == 0:
        return '00' * byte_len

    negative = n < 0
    abs_n = -n if negative else n

    result_bytes = []
    while abs_n > 0:
        result_bytes.append(abs_n & 0xff)
        abs_n >>= 8

    # Sign bit handling
    if result_bytes:
        if negative:
            if (result_bytes[-1] & 0x80) == 0:
                result_bytes[-1] |= 0x80
            else:
                result_bytes.append(0x80)
        else:
            if (result_bytes[-1] & 0x80) != 0:
                result_bytes.append(0x00)

    # Pad or truncate to requested length
    while len(result_bytes) < byte_len:
        result_bytes.append(0x00)
    result_bytes = result_bytes[:byte_len]

    return ''.join(f'{b:02x}' for b in result_bytes)


def _bin2num_int(hex_str: str) -> int:
    if not hex_str:
        return 0
    result_bytes = []
    for i in range(0, len(hex_str), 2):
        result_bytes.append(int(hex_str[i:i + 2], 16))
    if not result_bytes:
        return 0

    negative = (result_bytes[-1] & 0x80) != 0
    if negative:
        result_bytes[-1] &= 0x7f

    result = 0
    for i in range(len(result_bytes) - 1, -1, -1):
        result = (result << 8) | result_bytes[i]

    return -result if negative else result

File: packages/runar-sdk/src/anf-interpreter.ts
/**
 * Lightweight ANF interpreter for auto-computing state transitions.
 *
 * Given a compiled artifact's ANF IR, the current contract state, and
 * method arguments, this interpreter walks the ANF bindings and computes
 * the new state. It handles `update_prop` nodes to track state mutations,
 * while skipping on-chain-only operations like `check_preimage`,
 * `deserialize_state`, `get_state_script`, `add_output`, and `add_raw_output`.
 *
 * This enables the SDK to auto-compute `newState` for stateful contract
 * calls, so callers don't need to duplicate contract logic.
 */

import type {
  ANFProgram,
  ANFBinding,
  ANFValue,
} from 'runar-ir-schema';
import { Hash, Utils } from '@bsv/sdk';

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Compute the new state after executing a contract method.
 *
 * @param anf         The ANF IR from the compiled artifact.
 * @param methodName  The method to execute (must be a public method).
 * @param currentState  Current contract state (property name → value).
 * @param args        Method arguments (param name → value).
 * @returns The updated state (merged with currentState).
 */
export function computeNewState(
  anf: ANFProgram,
  methodName: string,
  currentState: Record<string, unknown>,
  args: Record<string, unknown>,
): Record<string, unknown> {
  // Find the method in ANF
  const method = anf.methods.find(
    (m) => m.name === methodName && m.isPublic,
  );
  if (!method) {
    throw new Error(
      `computeNewState: method '${methodName}' not found in ANF IR`,
    );
  }

  // Initialize the environment with property values and method params
  const env: Record<string, unknown> = {};

  // Load properties
  for (const prop of anf.properties) {
    env[prop.name] = currentState[prop.name] ?? prop.initialValue;
  }

  // Load method params (skip implicit ones injected by the compiler)
  const implicitParams = new Set([
    '_changePKH', '_changeAmount', '_newAmount', 'txPreimage',
  ]);
  for (const param of method.params) {
    if (implicitParams.has(param.name)) continue;
    if (param.name in args) {
      env[param.name] = args[param.name];
    }
  }

  // Track state mutations
  const stateDelta: Record<string, unknown> = {};

  // Walk bindings
  evalBindings(method.body, env, stateDelta, anf);

  return { ...currentState, ...stateDelta };
}

// ---------------------------------------------------------------------------
// Binding evaluation
// ---------------------------------------------------------------------------

function evalBindings(
  bindings: ANFBinding[],
  env: Record<string, unknown>,
  stateDelta: Record<string, unknown>,
  anf?: ANFProgram,
): void {
  for (const binding of bindings) {
    const val = evalValue(binding.value, env, stateDelta, anf);
    env[binding.name] = val;
  }
}

function evalValue(
  value: ANFValue,
  env: Record<string, unknown>,
  stateDelta: Record<string, unknown>,
  anf?: ANFProgram,
): unknown {
  switch (value.kind) {
    case 'load_param':
      return env[value.name];

    case 'load_prop':
      return env[value.name];

    case 'load_const': {
      const v = value.value;
      // Handle @ref: aliases (load_const with "@ref:targetName")
      if (typeof v === 'string' && v.startsWith('@ref:')) {
        return env[v.slice(5)];
      }
      return v;
    }

    case 'bin_op':
      return evalBinOp(
        value.op,
        env[value.left],
        env[value.right],
        value.result_type,
      );

    case 'unary_op':
      return evalUnaryOp(value.op, env[value.operand], value.result_type);

    case 'call':
      return evalCall(value.func, value.args.map((a) => env[a]));

    case 'method_call':
      return evalMethodCall(
        env,
        value.method,
        value.args.map((a: string) => env[a]),
        stateDelta,
        anf,
      );

    case 'if': {
      const cond = env[value.cond];
      const branch = isTruthy(cond) ? value.then : value.else;
      // Create a child env for the branch
      const childEnv = { ...env };
      evalBindings(branch, childEnv, stateDelta, anf);
      // Copy any new bindings back (the last binding is typically the branch result)
      Object.assign(env, childEnv);
      // Return the last binding's value from the branch
      if (branch.length > 0) {
        return childEnv[branch[branch.length - 1]!.name];
      }
      return undefined;
    }

    case 'loop': {
      const { count, body, iterVar } = value;
      let lastVal: unknown;
      for (let i = 0; i < count; i++) {
        env[iterVar] = BigInt(i);
        const loopEnv = { ...env };
        evalBindings(body, loopEnv, stateDelta, anf);
        // Copy loop bindings back
        Object.assign(env, loopEnv);
        if (body.length > 0) {
          lastVal = loopEnv[body[body.length - 1]!.name];
        }
      }
      return lastVal;
    }

    case 'assert': {
      // In simulation, we skip asserts (the on-chain script handles enforcement)
      return undefined;
    }

    case 'update_prop': {
      const newVal = env[value.value];
      env[value.name] = newVal;
      stateDelta[value.name] = newVal;
      return undefined;
    }

    case 'add_output': {
      // Extract implicit state changes from stateValues array.
      // stateValues[i] maps to the i-th mutable property (declaration order).
      if (anf && value.stateValues && value.stateValues.length > 0) {
        const mutableProps = anf.properties.filter((p) => !p.readonly);
        for (let i = 0; i < value.stateValues.length && i < mutableProps.length; i++) {
          const propName = mutableProps[i]!.name;
          const ref = value.stateValues[i]!;
          const newVal = env[ref];
          env[propName] = newVal;
          stateDelta[propName] = newVal;
        }
      }
      return undefined;
    }

    // On-chain-only operations — skip in simulation
    case 'check_preimage':
    case 'deserialize_state':
    case 'get_state_script':
    case 'add_raw_output':
      return undefined;

    default:
      return undefined;
  }
}

// ---------------------------------------------------------------------------
// Binary operations
// ---------------------------------------------------------------------------

function evalBinOp(
  op: string,
  left: unknown,
  right: unknown,
  resultType?: string,
): unknown {
  if (resultType === 'bytes' || (typeof left === 'string' && typeof right === 'string')) {
    return evalBytesBinOp(op, String(left ?? ''), String(right ?? ''));
  }

  const l = toBigInt(left);
  const r = toBigInt(right);

  switch (op) {
    case '+': return l + r;
    case '-': return l - r;
    case '*': return l * r;
    case '/': return r === 0n ? 0n : l / r;
    case '%': return r === 0n ? 0n : l % r;
    case '==': case '===': return l === r;
    case '!=': case '!==': return l !== r;
    case '<': return l < r;
    case '<=': return l <= r;
    case '>': return l > r;
    case '>=': return l >= r;
    case '&&': return isTruthy(left) && isTruthy(right);
    case '||': return isTruthy(left) || isTruthy(right);
    case '&': return l & r;
    case '|': return l | r;
    case '^': return l ^ r;
    case '<<': return l << r;
    case '>>': return l >> r;
    default: return 0n;
  }
}

function evalBytesBinOp(op: string, left: string, right: string): unknown {
  switch (op) {
    case '+':  // cat
      return left + right;
    case '==': case '===':
      return left === right;
    case '!=': case '!==':
      return left !== right;
    default:
      return '';
  }
}

// ---------------------------------------------------------------------------
// Unary operations
// ---------------------------------------------------------------------------

function evalUnaryOp(op: string, operand: unknown, resultType?: string): unknown {
  if (resultType === 'bytes') {
    // Bitwise NOT on bytes
    if (op === '~') {
      const hex = String(operand ?? '');
      const bytes = Utils.toArray(hex, 'hex');
      for (let i = 0; i < bytes.length; i++) bytes[i] = ~bytes[i]! & 0xff;
      return Utils.toHex(bytes);
    }
    return operand;
  }

  const val = toBigInt(operand);
  switch (op) {
    case '-': return -val;
    case '!': return !isTruthy(operand);
    case '~': return ~val;
    default: return val;
  }
}

// ---------------------------------------------------------------------------
// Built-in function calls
// ---------------------------------------------------------------------------

function evalCall(func: string, args: unknown[]): unknown {
  switch (func) {
    // Crypto — mock
    case 'checkSig': return true;
    case 'checkMultiSig': return true;
    case 'checkPreimage': return true;

    // Crypto — real hashes
    case 'sha256': return hashFn('sha256', args[0]);
    case 'hash256': return hashFn('hash256', args[0]);
    case 'hash160': return hashFn('hash160', args[0]);
    case 'ripemd160': return hashFn('ripemd160', args[0]);

    // Assert — skip (on-chain handles it)
    case 'assert': return undefined;

    // Byte operations
    case 'num2bin': {
      const n = toBigInt(args[0]);
      const len = Number(toBigInt(args[1]));
      return num2binHex(n, len);
    }
    case 'bin2num': {
      return bin2numBigInt(String(args[0] ?? ''));
    }
    case 'cat': {
      return String(args[0] ?? '') + String(args[1] ?? '');
    }
    case 'substr': {
      const hex = String(args[0] ?? '');
      const start = Number(toBigInt(args[1]));
      const len = Number(toBigInt(args[2]));
      return hex.slice(start * 2, (start + len) * 2);
    }
    case 'reverseBytes': {
      const hex = String(args[0] ?? '');
      const pairs: string[] = [];
      for (let i = 0; i < hex.length; i += 2) pairs.push(hex.slice(i, i + 2));
      return pairs.reverse().join('');
    }
    case 'len': {
      const hex = String(args[0] ?? '');
      return BigInt(hex.length / 2);
    }

    // Math builtins
    case 'abs': return toBigInt(args[0]) < 0n ? -toBigInt(args[0]) : toBigInt(args[0]);
    case 'min': return toBigInt(args[0]) < toBigInt(args[1]) ? toBigInt(args[0]) : toBigInt(args[1]);
    case 'max': return toBigInt(args[0]) > toBigInt(args[1]) ? toBigInt(args[0]) : toBigInt(args[1]);
    case 'within': {
      const x = toBigInt(args[0]);
      return x >= toBigInt(args[1]) && x < toBigInt(args[2]);
    }
    case 'safediv': {
      const d = toBigInt(args[1]);
      return d === 0n ? 0n : toBigInt(args[0]) / d;
    }
    case 'safemod': {
      const d = toBigInt(args[1]);
      return d === 0n ? 0n : toBigInt(args[0]) % d;
    }
    case 'clamp': {
      const v = toBigInt(args[0]);
      const lo = toBigInt(args[1]);
      const hi = toBigInt(args[2]);
      return v < lo ? lo : v > hi ? hi : v;
    }
    case 'sign': {
      const v = toBigInt(args[0]);
      return v > 0n ? 1n : v < 0n ? -1n : 0n;
    }
    case 'pow': {
      const base = toBigInt(args[0]);
      const exp = toBigInt(args[1]);
      if (exp < 0n) return 0n;
      let result = 1n;
      for (let i = 0n; i < exp; i++) result *= base;
      return result;
    }
    case 'sqrt': {
      const v = toBigInt(args[0]);
      if (v <= 0n) return 0n;
      let x = v;
      let y = (x + 1n) / 2n;
      while (y < x) { x = y; y = (x + v / x) / 2n; }
      return x;
    }
    case 'gcd': {
      let a = toBigInt(args[0]);
      let b = toBigInt(args[1]);
      if (a < 0n) a = -a;
      if (b < 0n) b = -b;
      while (b !== 0n) { const t = b; b = a % b; a = t; }
      return a;
    }
    case 'divmod': {
      const a = toBigInt(args[0]);
      const b = toBigInt(args[1]);
      if (b === 0n) return 0n;
      // Returns quotient; in ANF the second result is in a separate binding
      return a / b;
    }
    case 'log2': {
      const v = toBigInt(args[0]);
      if (v <= 0n) return 0n;
      let bits = 0n;
      let x = v;
      while (x > 1n) { x >>= 1n; bits++; }
      return bits;
    }
    case 'bool': return isTruthy(args[0]) ? 1n : 0n;
    case 'mulDiv': {
      return (toBigInt(args[0]) * toBigInt(args[1])) / toBigInt(args[2]);
    }
    case 'percentOf': {
      return (toBigInt(args[0]) * toBigInt(args[1])) / 10000n;
    }

    // Preimage intrinsics — return dummy values in simulation
    case 'extractOutputHash':
    case 'extractAmount':
      return '00'.repeat(32);

    default:
      return undefined;
  }
}

function evalMethodCall(
  callerEnv: Record<string, unknown>,
  methodName: string,
  args: unknown[],
  stateDelta: Record<string, unknown>,
  anf?: ANFProgram,
): unknown {
  // Private method calls appear in the ANF with their bodies available
  // in anf.methods. Execute the method body to compute its return value.
  if (anf) {
    const method = anf.methods.find(
      (m) => m.name === methodName && !m.isPublic,
    );
    if (method) {
      // Build env for the private method: copy property values for load_prop
      const methodEnv: Record<string, unknown> = {};
      for (const prop of anf.properties) {
        if (prop.name in callerEnv) {
          methodEnv[prop.name] = callerEnv[prop.name];
        }
      }

      // Map method params to passed args
      for (let i = 0; i < method.params.length && i < args.length; i++) {
        methodEnv[method.params[i]!.name] = args[i];
      }

      // Execute the method body — pass real stateDelta so update_prop
      // mutations in private methods are captured
      evalBindings(method.body, methodEnv, stateDelta, anf);

      // Propagate property changes back to the caller's env
      for (const prop of anf.properties) {
        if (prop.name in methodEnv) {
          callerEnv[prop.name] = methodEnv[prop.name];
        }
      }

      // Return the last binding's value (the method's return value)
      if (method.body.length > 0) {
        return methodEnv[method.body[method.body.length - 1]!.name];
      }
      return undefined;
    }
  }
  return undefined;
}

// ---------------------------------------------------------------------------
// Hash helpers
// ---------------------------------------------------------------------------

function hashFn(
  name: 'sha256' | 'hash256' | 'hash160' | 'ripemd160',
  input: unknown,
): string {
  const hex = String(input ?? '');
  const bytes = Utils.toArray(hex, 'hex');
  let result: number[];
  switch (name) {
    case 'sha256': result = Hash.sha256(bytes); break;
    case 'hash256': result = Hash.hash256(bytes); break;
    case 'hash160': result = Hash.hash160(bytes); break;
    case 'ripemd160': result = Hash.ripemd160(bytes); break;
  }
  return Utils.toHex(result);
}

// ---------------------------------------------------------------------------
// Numeric helpers
// ---------------------------------------------------------------------------

function toBigInt(v: unknown): bigint {
  if (typeof v === 'bigint') return v;
  if (typeof v === 'number') return BigInt(v);
  if (typeof v === 'boolean') return v ? 1n : 0n;
  if (typeof v === 'string') {
    // Handle "42n" format from JSON
    if (/^-?\d+n$/.test(v)) return BigInt(v.slice(0, -1));
    // Handle plain numeric strings
    if (/^-?\d+$/.test(v)) return BigInt(v);
    return 0n;
  }
  return 0n;
}

function isTruthy(v: unknown): boolean {
  if (typeof v === 'boolean') return v;
  if (typeof v === 'bigint') return v !== 0n;
  if (typeof v === 'number') return v !== 0;
  if (typeof v === 'string') return v !== '' && v !== '0' && v !== 'false';
  return false;
}

// ---------------------------------------------------------------------------
// Byte encoding helpers
// ---------------------------------------------------------------------------

function num2binHex(n: bigint, byteLen: number): string {
  if (n === 0n) return '00'.repeat(byteLen);

  const negative = n < 0n;
  let abs = negative ? -n : n;

  const bytes: number[] = [];
  while (abs > 0n) {
    bytes.push(Number(abs & 0xffn));
    abs >>= 8n;
  }

  // Sign bit handling: if MSB has sign bit set and number is positive,
  // or vice versa, add a padding byte
  if (bytes.length > 0) {
    if (negative) {
      if ((bytes[bytes.length - 1]! & 0x80) === 0) {
        bytes[bytes.length - 1]! |= 0x80;
      } else {
        bytes.push(0x80);
      }
    } else {
      if ((bytes[bytes.length - 1]! & 0x80) !== 0) {
        bytes.push(0x00);
      }
    }
  }

  // Pad or truncate to requested length
  while (bytes.length < byteLen) bytes.push(0x00);
  bytes.length = byteLen;

  return bytes.map((b) => b.toString(16).padStart(2, '0')).join('');
}

function bin2numBigInt(hex: string): bigint {
  if (!hex || hex.length === 0) return 0n;
  const bytes: number[] = [];
  for (let i = 0; i < hex.length; i += 2) {
    bytes.push(parseInt(hex.slice(i, i + 2), 16));
  }
  if (bytes.length === 0) return 0n;

  const negative = (bytes[bytes.length - 1]! & 0x80) !== 0;
  if (negative) {
    bytes[bytes.length - 1]! &= 0x7f;
  }

  let result = 0n;
  for (let i = bytes.length - 1; i >= 0; i--) {
    result = (result << 8n) | BigInt(bytes[i]!);
  }

  return negative ? -result : result;
}
